catalyst_voices_packages/catalyst_cardano_serialization/lib/src/transaction_output.dart

```diff
@@ -0,0 +1,363 @@
+/// a CBOR.
+sealed class Datum extends Equatable implements CborEncodable {
+  const Datum();
+}
+
+/// Class representing an optional datum in a Cardano transaction output.
+/// A [DatumOption] wraps a [Datum], which can either be a datum hash or direct
+/// data as inline datum.
+///
+/// This class is used to support additional features in transaction outputs
+/// that were introduced in Cardano after the Alonzo hard fork.
+final class DatumOption extends Datum {
+  /// The actual datum wrapped by this [DatumOption].
+  final Datum datum;
+
+  /// Constructor for the [DatumOption] class, which requires a [Datum].
+  const DatumOption(this.datum);
+
+  /// Factory constructor to create a [DatumOption] from a CBOR-encoded value.
+  /// The CBOR value is expected to be a list where the first element is the
+  /// type (either datum hash or data) and the second element is the actual
+  /// encoded datum.
+  factory DatumOption.fromCbor(CborValue value) {
+    if (value is CborList) {
+      final type = DatumOptionType.values[(value[0] as CborSmallInt).toInt()];
+      final datum = switch (type) {
+        DatumOptionType.datumHash => DatumHash.fromCbor(value[1]),
+        DatumOptionType.data => Data.fromCbor(value[1]),
+      };
+      return DatumOption(datum);
+    } else {
+      throw ArgumentError.value(
+        value,
+        'value',
+        'Expected DatumOption as CborList',
+      );
+    }
+  }
+
+  /// Converts the [DatumOption] to a CBOR-encoded value.
+  /// The CBOR list contains an index indicating the type and the encoded datum.
+  @override
+  CborValue toCbor() {
+    final index = switch (datum) {
+      DatumHash _ => DatumOptionType.datumHash,
+      Data _ => DatumOptionType.data,
+      _ => throw ArgumentError.value(datum, 'datum', 'Invalid datum type'),
+    };
+    return CborList([CborSmallInt(index.value), datum.toCbor()]);
+  }
+
+  @override
+  List<Object?> get props => [datum];
+}
+
+/// Class representing a datum hash in Cardano.
+/// A [DatumHash] is a type of [Datum] that contains a cryptographic hash, which
+/// refers to data stored off-chain.
+///
+/// This is useful when you need to reference data without including the actual
+/// data in the transaction output.
+final class DatumHash extends Datum {
+  // TODO(ilap): Add length constraint and/or implement proper hash type.
+  /// Byte array representing the cryptographic hash.
+  final Uint8List hash;
+
+  /// Constructor for the [DatumHash] class, which requires a hash.
+  const DatumHash(this.hash);
+
+  /// Converts the [DatumHash] to a CBOR-encoded value.
+  /// The CBOR value is a byte string representing the hash.
+  @override
+  CborValue toCbor() => CborBytes(hash);
+
+  @override
+  List<Object?> get props => [hash];
+
+  /// Factory constructor to create a [DatumHash] from a CBOR-encoded value.
+  /// The CBOR value is expected to be a byte string.
+  factory DatumHash.fromCbor(CborValue value) {
```

Comment:
For DatumHash the `fromCbor` constructor should be directly under `const DatumHash(this.hash)` constructor, then `toCbor` should follow and lastly there should be `props`. It's the convention that other files follow to make it easier to keep `fromCbor / toCbor` in sync.

Reply:
Please review other classes in this file and apply this convention 👍

Reply:
Done